### ISSUE TITLE
Plans for Dino + SAM2

### ISSUE THREAD

**Issue by lab176344** [autodistill/autodistill#171]
### Search before asking

- [X] I have searched the Autodistill [issues](https://github.com/autodistill/autodistill/issues) and found no similar feature requests.

### Description

I think combining DINO+SAM2 for video annotation might be a good base model. @capjamesg 

### Use case

Instance segmentation

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by capjamesg** [comment]
Hello! We already have a module that combines Grounding DINO / Florence-2 with SAM 2 ([autodistill-grounded-sam-2](https://github.com/autodistill/autodistill-grounded-sam-2)). It only supports image annotation at the moment, though. I think it would be cool for us to have a video annotation system! How are you thinking about making it?

**Comment by lab176344** [comment]
Ok good to know, we can extend the current autodistil-grounded-sam2 to take in video as an input, and per image sampled from video, additional to the masks we can add the tracking id to detection, i think tracking id column can be added to the yolo format as <class_id> <x_center> <y_center> <width> <height> <tracking_id> and also in coco_json in annotations

**Comment by capjamesg** [comment]
One notable limitation with SAM 2 for video is that you need to mark every object individually. Suppose you have a video of 10 shipping containers and only two are in frame at any given time. The first two you tag will be tracked, but the other eight will not be unless you mark them. Is there a good solution around this?

**Comment by lab176344** [comment]
yeah that is a good point, what if when we run florence + grounded dino and get all the bounding box first and look for masks closer to the boxes at each frame and update the sam predictor with new points. I don't know if the add_new_points from SAM2 supports that have to check. Also there might be some error due to the bbox check against mask

**Comment by lab176344** [comment]
I will do a POC, setting things up now. Will decide later how to integrate that with autodistill

**Comment by HanClinto** [comment]
I would be interested in helping on this -- using SAM 2 for auto-labeling (with tracking IDs) in video is an extremely relevant use-case for us.

**Comment by lab176344** [comment]
did not manage to start this yet, will be able to start it by the weekend i will create a new branch and PR we can collaborate

**Comment by HanClinto** [comment]
> One notable limitation with SAM 2 for video is that you need to mark every object individually. Suppose you have a video of 10 shipping containers and only two are in frame at any given time. The first two you tag will be tracked, but the other eight will not be unless you mark them. Is there a good solution around this?

I'm still mulling over this concern -- I think that what we might need to do is to run the prompt against every video frame (or every X video frames) and see if there are any _new_ detections of that object that weren't caught in previous frames. It will be a slow process, but I don't know of any other way around it -- no other model is set up to look at an entire video at once the way that SAM 2 is, and until we get prompt-driven models that have a similar "holistic" look at video, then we will continue running into this problem.

**Comment by lab176344** [comment]
> > One notable limitation with SAM 2 for video is that you need to mark every object individually. Suppose you have a video of 10 shipping containers and only two are in frame at any given time. The first two you tag will be tracked, but the other eight will not be unless you mark them. Is there a good solution around this?
> 
> I'm still mulling over this concern -- I think that what we might need to do is to run the prompt against every video frame (or every X video frames) and see if there are any _new_ detections of that object that weren't caught in previous frames. It will be a slow process, but I don't know of any other way around it -- no other model is set up to look at an entire video at once the way that SAM 2 is, and until we get prompt-driven models that have a similar "holistic" look at video, then we will continue running into this problem.

Yes, i think that is the only way i can also think of. Run grounded dino on all frames save it in a data store and for every frame processed by sam2 compare it against the datastore and reset the state with old and new bbox.

**Comment by lab176344** [comment]
Sorry took longer than expected as got busy with other things, I think I have a dirty implementation now which works. Will clean up and raise a PR against https://github.com/autodistill/autodistill-grounded-sam-2. Is that okay @capjamesg then we do video specific table in autodistill

**Comment by lab176344** [comment]
https://github.com/autodistill/autodistill-grounded-sam-2/pull/8 created a WIP PR here
